### PR TITLE
GlusterFS: Upgrade playbook

### PR DESCRIPTION
**Pull request by jarrpa**
This PR introduces a new playbook to perform a rolling upgrade of existing Gluster clusters on OpenShift. There were a few refactors and small tweaks required to facilitate this playbook.

Signed-off-by: Jose A. Rivera <jarrpa@redhat.com>

### CODE REVIEW

**Comment by jarrpa** [comment]
Dependent PR merged, so removing the WIP label. This is ready for review.

@michaelgugino PTAL?

File: playbooks/openshift-glusterfs/private/upgrade.yml [hunk 1]
```diff
@@ -0,0 +1,8 @@
+---
+- name: Upgrade GlusterFS
+  hosts: oo_first_master
+  tasks:
+  - name: Run glusterfs upgrade role
+    include_role:
```
Comment (michaelgugino) [inline playbooks/openshift-glusterfs/private/upgrade.yml]: import_role

File: roles/openshift_storage_glusterfs/tasks/restart_node.yml [hunk 2]
```diff
@@ -0,0 +1,33 @@
+---
```
Comment (michaelgugino) [inline roles/openshift_storage_glusterfs/tasks/restart_node.yml]: The name of this file is not great.

Should be recreate_pods.yml; restart_node.yml implies we're rebooting the host, or possibly restarting the kubelet.

File: roles/openshift_storage_glusterfs/tasks/restart_node.yml [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+- name: Unlabel GlusterFS nodes
```
Comment (michaelgugino) [inline roles/openshift_storage_glusterfs/tasks/restart_node.yml]: node

File: roles/openshift_storage_glusterfs/tasks/restart_node.yml [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+  oc_label:
+    name: "{{ restart_node.spec.nodeName }}"
+    kind: node
+    state: absent
+    labels: "{{ glusterfs_nodeselector | lib_utils_oo_dict_to_list_of_dict }}"
+
+- name: Unlabel GlusterFS nodes (old label)
```
Comment (michaelgugino) [inline roles/openshift_storage_glusterfs/tasks/restart_node.yml]: node

File: roles/openshift_storage_glusterfs/tasks/restart_node.yml [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+  oc_label:
+    name: "{{ restart_node.spec.nodeName }}"
+    kind: node
+    state: absent
+    labels: [ { 'key': 'storagenode', 'value': '' } ]
+
```
Comment (michaelgugino) [inline roles/openshift_storage_glusterfs/tasks/restart_node.yml]: Should add a comment why we're doing this here so it's more readable.

File: roles/openshift_storage_glusterfs/tasks/wait_for_pods.yml [hunk 6]
```diff
@@ -1,14 +1,19 @@
 ---
+- set_fact:
```
Comment (michaelgugino) [inline roles/openshift_storage_glusterfs/tasks/wait_for_pods.yml]: remove this set_fact.

File: roles/openshift_storage_glusterfs/tasks/wait_for_pods.yml [hunk 7]
```diff
@@ -1,14 +1,19 @@
...
     state: list
     selector: "glusterfs={{ glusterfs_name }}-pod"
-  register: glusterfs_pods
+  register: glusterfs_pods_wait
   until:
-  - "glusterfs_pods.results.results[0]['items'] | count > 0"
+  - "glusterfs_pods_wait.results.results[0]['items'] | count > 0"
   # There must be as many pods with 'Ready' staus  True as there are nodes expecting those pods
-  - "glusterfs_pods.results.results[0]['items'] | lib_utils_oo_collect(attribute='status.conditions') | lib_utils_oo_collect(attribute='status', filters={'type': 'Ready'}) | map('bool') | select | list | count == glusterfs_nodes | count"
+  - "glusterfs_pods_wait.results.results[0]['items'] | lib_utils_oo_collect(attribute='status.conditions') | lib_utils_oo_collect(attribute='status', filters={'type': 'Ready'}) | map('bool') | select | list | count == glusterfs_count | int"
```
Comment (michaelgugino) [inline roles/openshift_storage_glusterfs/tasks/wait_for_pods.yml]:
```yaml
count == l_glusterfs_count
...
vars:
  l_glusterfs_count: "{{ glusterfs_count | default(glusterfs_nodes | count) | int }}"
```

File: roles/openshift_storage_glusterfs/tasks/restart_node.yml [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+- set_fact:
```
Comment (michaelgugino) [inline roles/openshift_storage_glusterfs/tasks/restart_node.yml]: Remove this set_fact

File: roles/openshift_storage_glusterfs/tasks/restart_node.yml [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+    glusterfs_count: "{{ glusterfs_count | int - 1 }}"
+
+- import_tasks: wait_for_pods.yml
```
Comment (michaelgugino) [inline roles/openshift_storage_glusterfs/tasks/restart_node.yml]: add:
```yaml
vars:
  glusterfs_count: "{{ glusterfs_nodes | count - 1 }}"
```

File: roles/openshift_storage_glusterfs/tasks/restart_node.yml [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+
+- name: Label GlusterFS nodes
+  oc_label:
+    name: "{{ restart_node.spec.nodeName }}"
+    kind: node
+    state: add
+    labels: "{{ glusterfs_nodeselector | lib_utils_oo_dict_to_list_of_dict }}"
+
+- set_fact:
```
Comment (michaelgugino) [inline roles/openshift_storage_glusterfs/tasks/restart_node.yml]: Remove this.

File: playbooks/openshift-glusterfs/upgrade.yml [hunk 11]
```diff
@@ -0,0 +1,4 @@
+---
+- import_playbook: ../init/main.yml
```
[the remark this answers is not in the document]
Comment (jarrpa) [reply playbooks/openshift-glusterfs/upgrade.yml]: I don't know how to do that, but I'll copy the values from config.yml and see if that looks good to you.

**Comment by jarrpa** [comment]
@michaelgugino Comments addressed. Hopefully correctly. :)

File: playbooks/openshift-glusterfs/upgrade.yml [hunk 12]
```diff
@@ -0,0 +1,10 @@
...
+  vars:
+    l_init_fact_hosts: "oo_masters_to_config:oo_glusterfs_to_config"
+    l_openshift_version_set_hosts: "oo_masters_to_config:!oo_first_master"
+    l_sanity_check_hosts: "{{ groups['oo_masters_to_config'] | union(groups['oo_glusterfs_to_config']) }}"
+    l_install_base_packages: True
```
Comment (michaelgugino) [inline playbooks/openshift-glusterfs/upgrade.yml]: Set this to False, we don't need to reinstall base packages.

File: playbooks/openshift-glusterfs/upgrade.yml [hunk 13]
```diff
@@ -0,0 +1,10 @@
...
+    l_base_packages_hosts: "oo_glusterfs_to_config"
```
Comment (michaelgugino) [inline playbooks/openshift-glusterfs/upgrade.yml]: "all:!all"

**Comment by jarrpa** [comment]
@michaelgugino Comments addressed.

**Pull request merged**